Add `filter` feature for `var_names`

AlexAndorra opened the pull request:
## Description
- Currently, users can select variables entering `az.summary` and other plotting/diagnostics functions with the `var_names` kwarg. 
This selection must however be exact, meaning that if you have variables with similar names in your trace, you have to specify them all: `az.summary(idata, var_names=["alpha", "beta1", "beta2", "beta3"])`, which can be cumbersome for big models.

- This PR introduces the `filter_like` kwarg to allow users to select _all_ variables containing the given word. For instance, `az.summary(idata, var_names=["alpha", "beta"], filter_like=True)` would give the same result as above. This also works to _exclude_ variables with similar names: `az.summary(idata, var_names=["~beta"], filter_like=True)`

- `filter_like` defaults to `False` to ensure backwards-compatibility.

Before adding these changes to the changelog, I had pending questions:
1. Should we add `filter_like` to the other functionalities that use `var_names`, and not only to `summary`?
2. Should we include new tests for this feature?
- Bonus points: the new code may be too convoluted at times, so please tell me if you see a simpler, more pythonic way to do things.

I'm of course available for any changes, and thanks a lot in advances for your reviews 🖖

OriolAbril commented on arviz/stats/stats.py:
```diff
@@ -949,6 +950,9 @@ def summary(
         Names of variables to include in summary
+    filter_like: bool, default=False
+        Whether to interpret var_names as substrings of the real variables names. À la
+        `pandas.filter("like")`.
```
Does pandas.filter accept regular expressions?

AlexAndorra replied:
Yes, if you use `regex` as kwarg instead of `like`

OriolAbril replied:
Ohhh, I see now

OriolAbril commented on arviz/utils.py:
```diff
@@ -50,14 +51,26 @@ def _var_names(var_names, data):
+                for var in excluded_vars:
+                    if var not in all_vars:
+                        real_vars = [real_var for real_var in all_vars if var in real_var]
+                        excluded_vars.extend(real_vars)
+                        excluded_vars.remove(var)
```
what would happen if one of the `real_vars` was the same as `var`? i.e. `theta` in eight schools -> exlcude `theta` and `theta_t`

AlexAndorra replied:
Mmmh I don't think it's possible because `real_var` is in `all_vars`, but `var` is not in `all_vars`. Is that your question?

AlexAndorra replied:
Are we on the same page here @OriolAbril ?

AlexAndorra commented:
Thanks for the review @OriolAbril ! 

- I implemented the regex feature, extended the `filter` kwarg to all functions taking `var_names` as arg, and added tests for this new feature in `test_utils.py`.

- I also updated the examples in the docstrings when useful and deleted the white space between args' names and the colon in docstrings (as discussed in another PR).

- Here is how this works now: imagine that `all_vars=["alpha", "beta1", "beta2", "p1", "p2", "phi"]`. Users will now be able to:
    - Filter by partial naming: `az.summary(idata, var_names=["alpha", "beta"], filter="like") --> ["alpha", "beta1", "beta2"]`
    - `az.summary(idata, var_names=["~beta"], filter="like") --> ["alpha", "p1", "p2", "phi"]`
    - Filter by regular expressions: `az.summary(idata, var_names=["^bet"], filter="regex") --> ["beta1", "beta2"]`
    - `az.summary(idata, var_names=["~p[0-9]+"], filter="regex") --> ["alpha", "beta1", "beta2", "phi"]`

Tell me if you see other necessary changes, otherwise I hope this is ready to merge 🖖

ahartikainen commented on arviz/plots/autocorrplot.py:
```diff
@@ -30,18 +31,24 @@ def plot_autocorr(
+    filter: Union[None, "like", "regex"], optional, default=None
+        If `None` (default), interpret var_names as the real variables names. If "like",
+        interpret var_names as substrings of the real variables names. If "regex",
+        interpret var_names as regular expressions on the real variables names. À la
```
These special letters are nice until they hit some lib that handles only ascii etc (some versions of pip?; windows sometimes)

AlexAndorra replied:
Good catch! You're talking about "À", right?

ahartikainen replied:
Yes

OriolAbril commented on arviz/plots/autocorrplot.py:
```diff
@@ -30,18 +31,24 @@ def plot_autocorr(
+        Variables to be plotted, if None all variable are plotted. Prefix the
+        variables by `~` when you want to exclude them from the plot. Vector-value
+        stochastics are handled automatically.
+    filter: Union[None, "like", "regex"], optional, default=None
```
it should use docstring formatting, not type hints (type hints go in the function definition). In this case it would be:

    filter: {None, "like", "regex"}, optional

See references in https://github.com/arviz-devs/arviz/blob/master/CONTRIBUTING.md#docstring-formatting-and-type-hints

AlexAndorra replied:
My bad, will update, thanks 👌

ahartikainen commented on arviz/utils.py:
```diff
@@ -50,14 +54,34 @@ def _var_names(var_names, data):
-
-        existent_vars = np.isin(var_names, all_vars)
-        if not np.all(existent_vars):
+            if (filter == "like") or (filter == "regex"):
```
filter in ("like, "regex")

We should also call `str(filter).lower()` before this line

AlexAndorra replied:
You sound just like pylint 😜

aloctavodia commented on arviz/plots/autocorrplot.py:
```diff
@@ -88,7 +95,7 @@ def plot_autocorr(
         :context: close-figs
 
-        >>> az.plot_autocorr(data, var_names=['mu', 'tau'], combined=True)
+        >>> az.plot_autocorr(data, var_names=['~theta'], filter="like", combined=True)
```
The tittle should be different like     "Combine chains collapsing by variable and negate variable names".
setting `filter="like"`is not necessary to negate a variable, right?

AlexAndorra replied:
- You're right about the title, will do.
- Indeed, it `filter` is not necessary to negate, it was just to show an example of how to use it, to vary examples. I will use "thet" instead of the full variable name, to make `filter` necessary.

ahartikainen commented on arviz/utils.py:
```diff
@@ -50,14 +54,34 @@ def _var_names(var_names, data):
+                                real_var for real_var in all_vars if re.search(pattern, real_var)
+                            ]
+                        excluded_vars.extend(real_vars)
+                        excluded_vars.remove(pattern)
```
This will skip iterations. If needed, copy first the loop variable

AlexAndorra replied:
I'm not sure I understand -- can you expand a little please?

ahartikainen replied:
Actually this can blow-up too (infinite iteration). I will tell you when on computer.

(Never edit list while iterating over it)

aloctavodia commented on arviz/plots/essplot.py:
```diff
@@ -125,7 +132,7 @@ def plot_ess(
 
         >>> az.plot_ess(
-        ...     idata, kind="quantile", var_names=["mu", "theta"], coords=coords
+        ...     idata, kind="quantile", var_names=['~theta'], filter="like", coords=coords
```
Idem previous comment

OriolAbril commented on arviz/utils.py:
```diff
@@ -50,14 +54,34 @@ def _var_names(var_names, data):
+                            real_vars = [
+                                real_var for real_var in all_vars if re.search(pattern, real_var)
+                            ]
+                        excluded_vars.extend(real_vars)
+                        excluded_vars.remove(pattern)
```
Back to these two lines. I will try to better define my question, which could then be added as a test (I think). Imagine we have `ds.data_vars = ["theta", "theta_t"]` and we call a given function with `az.xyz(..., var_names="~theta", filter="like")`. We then have that `excluded_vars = ["theta"]` and in the pattern loop: `real_vars=["theta", "theta_t"]` which means that right after extending, `excluded_vars = ["theta", "theta", "theta_t"]` and then the `.remove` deletes the first occurrence of theta. Eventually, as `.remove` only deletes the first occurrence it does work, but it feels very convoluted. I would remove `pattern` at the beginning of the iteration and leave the extend at the end, this would make the operation feel more like _substituting the pattern for the list of matches_ and would not require people reading the code to know that `.remove` deletes only the first occurrence to be able to understand what is happening.

Just realized about the `if pattern not in all_vars` which would make all of the above not happen if `pattern == "theta"` right? Sorry I have not been able to try the code and my _brain python_ tends to miss important details.

AlexAndorra replied:
Yes, you're right, in that case your example would not happen because `theta` is in `all_vars`. However, in this case, as the loop is not executed, only `theta` will be excluded and `theta_t` will still be there.
So the solution to you is to remove the `if pattern not in all_vars` condition, right?

On your second point, I'll indeed move the `.remove` -- or just get rid of it entirely (see Ari's comment above)

OriolAbril commented on arviz/utils.py:
```diff
@@ -50,14 +54,34 @@ def _var_names(var_names, data):
-        if not np.all(existent_vars):
+            if (filter == "like") or (filter == "regex"):
+                for pattern in excluded_vars:
+                    if pattern not in all_vars:
```
I _think_ this should not be here (see below too) because it should be valid to use the name itself even if we are using filter, the important check is the check above of `var not in all_vars` when creating `excluded_vars`

OriolAbril commented on arviz/tests/base_tests/test_utils.py:
```diff
@@ -64,6 +57,36 @@ def test_var_names_key_error(data):
+)
+def test_var_names_filter(var_names_expected):
+    """Test var_name filter with partial naming or regular expressions."""
+    data = from_dict(
+        posterior={
+            "alpha": np.random.randn(10),
+            "beta1": np.random.randn(10),
+            "beta2": np.random.randn(10),
+            "p1": np.random.randn(10),
+            "p2": np.random.randn(10),
+            "phi": np.random.randn(10),
+        }
+    ).posterior
```
I think this could use `dict_to_dataset` to avoid doing the `.posterior` and get the dataset straight away.

OriolAbril commented on arviz/tests/base_tests/test_utils.py:
```diff
@@ -64,6 +57,36 @@ def test_var_names_key_error(data):
 
 
+@pytest.mark.parametrize(
+    "var_names_expected",
```
given that this is `var_names`, `expected` and `filter`, I would name it `args`

AlexAndorra replied:
I opted for `var_args` -- `args` seemed too generic.

AlexAndorra commented:
- Addressed all comments
- Closed issue #1163 cc @canyon289

AlexAndorra commented:
Renamed the `filter` arg to `filter_vars`, both to keep pylint happy and to have a more explicit API. Should be ready to merge now 🎉
